src/FSMCIO.cpp

```diff
@@ -0,0 +1,172 @@
+    hsram.Init.PageSize = FSMC_PAGE_SIZE_NONE;
+    hsram.Init.ContinuousClock = FSMC_CONTINUOUS_CLOCK_SYNC_ONLY;
+
+    HAL_SRAM_Init(&hsram, &sram_timing, &sram_timing_write);
+}
+
+static FSMCIO *theInstance;
+
+FSMCIO::FSMCIO(uint32_t flags, PinNumber wr, PinNumber rd)
+{
+    ZERO(hsram);
+    ZERO(hdma);
+
+    theInstance = this;
+
+    GPIO_InitTypeDef gpio_init_structure;
+
+    gpio_init_structure.Mode = GPIO_MODE_AF_PP;
+    gpio_init_structure.Pull = GPIO_PULLUP;
+    gpio_init_structure.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+    gpio_init_structure.Alternate = GPIO_AF12_FSMC;
+
+    __HAL_RCC_FSMC_CLK_ENABLE();
+
+    int hasRead = rd != 0xff;
+
+    // 100+ pin version
+    if (wr == PD_5)
+    {
+        if (hasRead && rd != PD_4)
+            oops();
+
+        __HAL_RCC_GPIOD_CLK_ENABLE();
+        __HAL_RCC_GPIOE_CLK_ENABLE();
+        __HAL_RCC_GPIOF_CLK_ENABLE();
+        gpio_init_structure.Pin = GPIO_PIN_0 | GPIO_PIN_1 | GPIO_PIN_14 | GPIO_PIN_15 | GPIO_PIN_5;
+        if (rd == PD_4)
+            gpio_init_structure.Pin |= GPIO_PIN_4;
+        HAL_GPIO_Init(GPIOD, &gpio_init_structure);
+
+        gpio_init_structure.Pin = GPIO_PIN_7 | GPIO_PIN_8 | GPIO_PIN_9 | GPIO_PIN_10;
+        HAL_GPIO_Init(GPIOE, &gpio_init_structure);
+    }
+    else if (wr == PC_2 || wr == PD_2)
+    {
+        if (hasRead && rd != PC_5)
+            oops();
+
+        __HAL_RCC_GPIOA_CLK_ENABLE();
+        __HAL_RCC_GPIOB_CLK_ENABLE();
+        __HAL_RCC_GPIOC_CLK_ENABLE();
+
+        gpio_init_structure.Pin = GPIO_PIN_2 | GPIO_PIN_3 | GPIO_PIN_4 | GPIO_PIN_5;
+        HAL_GPIO_Init(GPIOA, &gpio_init_structure);
+
+        gpio_init_structure.Pin = GPIO_PIN_14;
+        HAL_GPIO_Init(GPIOB, &gpio_init_structure);
+
+        gpio_init_structure.Pin = GPIO_PIN_6 | GPIO_PIN_11 | GPIO_PIN_12;
+        if (wr == PC_2)
+            gpio_init_structure.Pin |= GPIO_PIN_2;
+        if (rd == PC_5)
+            gpio_init_structure.Pin |= GPIO_PIN_5;
+        HAL_GPIO_Init(GPIOC, &gpio_init_structure);
+
+        if (wr == PD_2)
+        {
+            __HAL_RCC_GPIOD_CLK_ENABLE();
+            gpio_init_structure.Pin = GPIO_PIN_2;
+            HAL_GPIO_Init(GPIOD, &gpio_init_structure);
+        }
+    }
+    else
+    {
+        oops();
+    }
+
+    FMC_BANK1_Init(flags, hsram);
+    dma_init(0, DMA_TX, &hdma, 0);
+    __HAL_LINKDMA(&hsram, hdma, hdma);
+}
+
+#define FSMC_DATA *((volatile uint8_t *)0x60000000)
+
+void FSMCIO::send(const void *txBuffer, uint32_t txSize)
+{
+    uint8_t *ptr = (uint8_t *)txBuffer;
+    while (txSize--)
+        FSMC_DATA = *ptr++;
+}
+
+static void HAL_SRAM_DMA_XferCpltCallback(DMA_HandleTypeDef *hdma)
+{
+    UNUSED(hdma);
+    theInstance->doneHandler(theInstance->handlerArg);
+}
+
+static void HAL_SRAM_DMA_XferErrorCallback(DMA_HandleTypeDef *hdma)
+{
+    UNUSED(hdma);
+    oops();
+}
+
+void FSMCIO::startSend(const void *txBuffer, uint32_t txSize, PVoidCallback doneHandler,
```

Comment:
Ooooh whats the purpose of `handlerArg`?

Reply:
it's passed to the handler, when the handler is called; C-style, this is the same as current SPI